peteonrails/add zfs tutorial (#1174)
clearlinux/clear-linux-documentation: source/tutorials/zfs.rst

#.  Check whether you have an LTS or native kernel:

    .. code-block:: bash
  
       uname -r

    a. If 'native' appears in the kernel name, then install a native
       kernel with DKMS support:

       .. code-block:: bash

          sudo swupd bundle-add kernel-native-dkms

    #. If 'lts' appears in the kernel name, then install the latest LTS
       kernel with DKMS support:

       .. code-block:: bash

          sudo swupd bundle-add kernel-lts-dkms

#. Reboot and make sure you can start the new kernel.

Bundles
=======

Before you install OpenZFS, install these bundles.

.. code-block:: bash

    sudo swupd bundle-add os-core-dev devpkg-openssl devpkg-util-linux

Install
*******

Clone OpenZFS code
==================

In this section, you download the source code directly from the `ZFS on Linux repository`_.

.. note::
   
   OpenZFS will not run from this location. We copy the source code into a DKMS directory in the following steps. 

#. Create a directory. In this example, we use :file:`/opt/src/zfs`, 
   but you can choose any workspace you like. 

   .. code-block:: bash

      mkdir -p /opt/src/zfs

#. Clone the repository.

   .. code-block:: bash

>       git clone https://github.com/openzfs/zfs.git /opt/src/zfs

Reviewer: Agreed

Remember where you clone the git repository because you will need it
when you upgrade ZFS. Do not delete the source location when you have completed this tutorial. You will need it later.

Compile the module
==================

We will build the module using DKMS. This will enable us to keep the module up to date as new kernels are released in the future.

#. Check out the tagged version that you plan to use. As of the time of this
   writing, the latest release tag is `0.8.4`:

   .. code-block:: bash

      cd /opt/src/zfs
      git checkout 0.8.4

#. Copy the source code into `/usr/src/zfs-0.8.4`. This exposes the source
   code to DKMS. We will build the code from the new location:

   .. code-block:: bash

      sudo cp -Rv /opt/src/zfs /usr/src/zfs-0.8.4
      cd /usr/src/zfs-0.8.4

#. The ZFS distribution provides a script to build a suitable dkms.conf file.
   Build dkms.conf and install it into the DKMS tree.

   .. code-block:: bash

>       cd /usr/src/zfs-0.8.4

Reviewer 2: This step is duplicte with the `cd` above. One could be eliminated.

>       scripts/dkms.mkconf -n zfs -v 0.8.4 -f dkms.conf

Reviewer 2: ```suggestion
      sudo scripts/dkms.mkconf -n zfs -v 0.8.4 -f dkms.conf
```

            |--+ zpool.d/
            |--zfs-functions
      |--+ include/libzfs/ [contents omitted]
      |--+ lib/
      |--+ libexec/
         |--+ zfs/
            |-- zpool.d/
            |-- zed.d
      |--+ share/zfs/ [contents omitted]
      |--+ sbin/
         |--fsck.zfs
         |--zpool
         |--zdb
         |--zed
         |--zfs
         |--zhack
         |--zinject
         |--zpool
         |--ztest
         |--zstreamdump
      |--+ src/
         |--+ zfs-0.8.4/
         |--+ spl-0.8.4/

#. Load the new kernel module:

   .. code-block: bash

      sudo modprobe zfs

Set up systemd
==============

You should now have these unit files available. 

.. code-block:: console

   zfs-env-bootfs.service
   zfs-zed.service
   zfs-import-cache.service
   zfs-import-scan.service
   zfs-mount.service
   zfs-share.service
   zfs-volume-wait.service
  

If you want to use ZFS automatic zpool import and filesystem
mount services, link the systemd.unit files into /etc and enable them.

.. code-block:: bash

   sudo ln -s ./etc/systemd/system/zfs-import-cache.service /etc/systemd/system/
   sudo ln -s ./etc/systemd/system/zfs-mount.service /etc/systemd/system/

>    systemctl enable zfs-import-cache

Reviewer 2: ```suggestion
   sudo systemctl enable zfs-import-cache
```

same suggestion for all the `systemctl` commands below.

   systemctl enable zfs-import.target

   systemctl enable zfs-mount
   systemctl enable zfs.target

> OpenZFS requires you to explicitly install and enable the services you want. 

Reviewer 2: I suggest presenting this information earlier, towards the top of this sections header.

If you want to use other ZFS service units, symlink them similarly to the 
example shown above.

> Load the kernel module at boot

Reviewer 2: Is this only needed if a device in the boot path is using ZFS or for *any* ZFS filesystems to work *at all*? In other words, does it work automagically if the `mount` command is used at run time? If so, I suggest clarifying when a user would want to do this.

Author: @mvincerx @puneetse The kernel modules must be loaded before any ZFS filesystems are mounted. If the modules are not loaded, then the `mount` command will fail with an error (I think it says: "ZFS modules not loaded.")

So, how about this: "OpenZFS kernel modules must be loaded before any OpenZFS filesystems are mounted. For convenience, load the kernel modules at boot."

Reviewer 3: @puneetse. I want to merge this PR. Do you have any other feedback on my latest push?

cannot be distributed in binary form alongside the kernel.

Some distributions of Linux have a different opinion of
CDDL / GPL compatibility -- most notably Canonical's Ubuntu -- and
argue that a binary kernel module can be distributed alongside
Linux without infringement. Clear Linux does not share that view.
Therefore, zfs.ko must be built by the user under Clear Linux.

It is worth understanding the `argument against shipping zfs.ko along
with Linux`_ in addition to `the argument in favor`_.

If you follow this tutorial and build zfs.ko on your system,
*you should not redistribute that work in binary form*.

Note that there is no issue with distributing ZFS in
*source* form alongside Linux. According to those who argue
both against and in favor of binary distribution, the CDDL and
GPL are not incompatible with respect to redistribution of
source code.

There is nothing stopping a user from downloading the ZFS drivers,
compiling them, then installing them for their own use. This is no
different from downloading and using proprietary drivers like the Nvidia
Graphics or Broadcom wireless drivers.

Since the binaries for ZFS cannot be distributed with the

> Linux kernel binaries, you'll have some work to  ensure compatbility

Reviewer: ```suggestion
Linux kernel binaries, you'll have some work to ensure compatibility
```

when new kernels are released.

This is not always trivial, since new kernel release candidates will
invariably break *something* in ZFS. The ZFS-on-Linux team is
very good about catching up quickly, but on occasion you will find

> a lag between a new kernel release and a supported ZFS driver.

Reviewer: We may want to recommend using an LTS kernel here, to reduce the likelihood of breaking changes between the kernel and ZFS.

Reviewer 3: 👍

Reviewer 3: We may want to add that "users should test releases before deploying."

Author: I added some language that conveys these warnings.

Prerequisites
*************

To build ZFS, you need:

DKMS kernel
===========

This tutorial assumes you are using a DKMS kernel. You should read and
understand how to :ref:`kernel-modules-dkms`.

If you do not currently use a DKMS kernel, install it with the following steps.

> Check whether you have an lts or native kernel:

Reviewer: ```suggestion
Check whether you have an LTS or native kernel:
```

.. code-block:: bash

    uname -r

If 'native' appears in the kernel name, then install a native kernel with DKMS support:

.. code-block:: bash

    sudo swupd bundle-add kernel-native-dkms

> If 'lts' appears in the kernel name, then install the latest TLS kernel with DKMS support:

Reviewer: ```suggestion
If 'lts' appears in the kernel name, then install the latest LTS kernel with DKMS support:
```

.. code-block:: bash

    sudo swupd bundle-add kernel-lts-dkms

Reboot and make sure you can start the new kernel.

Bundles
=======

Install these build tools before you install ZFS.

If you are using a native kernel:

.. code-block:: bash

    sudo swupd bundle-add linux-dev

If you are using an LTS kernel:

.. code-block:: bash

    sudo swupd bundle-add linux-lts-dev

> 

Reviewer: The `-dev` bundles are automatically included by the corresponding `-dkms` bundles, so they don't need to be added explicitly here.

> Finally, install need these bundles, no matter which kernel you are using:

Reviewer: ```suggestion
Finally, install these bundles, no matter which kernel you are using:
```

.. code-block:: bash

    sudo swupd bundle-add os-core-dev devpkg-openssl devpkg-util-linux

Install
*******

Get the ZFS code
================

The ZFS codebase moves quickly, just like the kernel codebase. Therefore

> it's best that you get the source code from the ZFS on Linux repository,

Reviewer: Maybe make "ZFS on Linux" a hyperlink?

hosted on Github.

Fetch the ZFS repository via git:

.. code-block:: bash

>    git clone https://github.com/openzfs/zfs.git /opt/src/zfs

Reviewer: You'll need a `mkdir` here.

Author: Oops! Thank you. Added.

> Remember where you check-out the git repository, because you will need it

Reviewer: ```suggestion
Remember where you clone the git repository, because you will need it
```

This will install the zfs userspace tools to:
::

   + /usr/local/
   |--+ bin/
      |--zvol_wait
      |--zgenhostid
      |--raidz_test
   |--+ etc/
      |--+ zfs/
         |--* zed.d/
         |--+ zpool.d/
         |--zfs-functions
   |--+ include/libzfs/ [contents omitted]
   |--+ lib/
   |--+ libexec/
      |--+ zfs/
         |-- zpool.d/
         |-- zed.d
   |--+ share/zfs/ [contents omitted]
   |--+ sbin/
      |--fsck.zfs
      |--zpool
      |--zdb
      |--zed
      |--zfs
      |--zhack
      |--zinject
      |--zpool
      |--ztest
      |--zstreamdump
   |--+ src/
      |--+ zfs-0.8.4/
      |--+ spl-0.8.4/

Load the new kernel module:

.. code-block: bash

   sudo modprobe zfs

Set up systemd
==============

You should now have these unit files available:

```
zfs-env-bootfs.service
zfs-zed.service
zfs-import-cache.service
zfs-import-scan.service
zfs-mount.service
zfs-share.service
zfs-volume-wait.service
```

If you want to use ZFS automatic zpool import and filesystem
mount services, copy the systemd.unit files into /etc and enable them:

.. code-block:: bash

   sudo cp ./etc/systemd/system/*.service /etc/systemd/system/

>    sudo cp ./contrib/dracut/90zfs/zfs-env-bootfs.service /etc/systemd/system/

Reviewer: Might be better to symlink these?

Author: I updated to symlink just the two services we plan to use into /etc/systemd/system, and added a sentence explaining that the user should do the same for any other service units they plan to use.

   systemctl enable zfs-import-cache
   systemctl enable zfs-import.target

   systemctl enable zfs-mount
   systemctl enable zfs.target

If you prefer to use the zfs-mount-generator instead of zfs-mount, refer to the
Arch Linux Guide's `ZFS Mount Generator`_ section for details. 

Load the kernel module at boot
==============================

The ZFS module will not load automatically at boot. Load the zfs.ko module

> at boot time by with systemd.

Reviewer: ```suggestion
at boot time with systemd.
```

First, allow unsigned modules:

    echo "module.sig_unenforce" | sudo tee /etc/kernel/cmdline.d/allow-unsigned-modules.conf

> 

Reviewer: This is automatically provided by the `-dkms` bundles.

Author: Removed.

Systemd uses the `/etc/modules-load.d/` directory to load out-of-tree kernel modules.
Make sure that the directory exists:

    sudo mkdir -p /etc/modules-load.d

Create the configuration file:

>     sudo echo "zfs" | sudo tee /etc/modules-load.d/01-zfs.conf

Reviewer: ```suggestion
    echo "zfs" | sudo tee /etc/modules-load.d/01-zfs.conf
```

Reboot -- zfs.ko should be loaded automatically.

Caution
*******
When the Clear Linux kernel is upgraded, DKMS will attempt to rebuild your
zfs module for the new kernel. If you boot a new kernel and cannot find
your zpools:

- DKMS may not have rebuilt the module
- DKMS may not have autoinstalled the module
- The new kernel might introduce breaking changes that prevent zfs from compiling

To fix this situation, you may have to recompile zfs.ko with the new kernel code.
ZFS *might* not compile at all with the new kernel.

**So, be sure you don't put anything on that ZFS pool that you would need
in order to rebuild kernel modules.**

> 

Reviewer: Should add that you can check the results of dkms like this:
```
systemctl status dkms-new-kernel.service
```
or by checking the journal.

Author: Added

   systemctl enable zfs-mount
   systemctl enable zfs.target

OpenZFS requires you to explicitly install and enable the services you want. 
If you want to use other ZFS service units, symlink them similarly to the 
example shown above.

Load the kernel module at boot
==============================

The OpenZFS module will not load automatically at boot. Load the zfs.ko module at boot time with systemd.

#. Systemd uses the `/etc/modules-load.d/` directory to load out-of-tree
   kernel modules. Make sure that the directory exists:

   .. code-block:: bash

      sudo mkdir -p /etc/modules-load.d

#. Create the configuration file:

   .. code-block:: bash

      echo "zfs" | sudo tee /etc/modules-load.d/01-zfs.conf

#. Reboot your system. zfs.ko should be loaded automatically.

.. CAUTION::

   When the |CL| kernel is upgraded, DKMS will attempt to rebuild your OpenZFS module for the new kernel. 

   - DKMS may not have rebuilt the module
   - DKMS may not have autoinstalled the module
   - The new kernel might introduce breaking changes that prevent zfs from 
     compiling

> To fix this situation, you may have to recompile zfs.ko with the new kernel code.OpenZFS *might* not compile at all with the new kernel.

Reviewer: ```suggestion
To fix this situation, you may have to recompile zfs.ko with the new kernel code. OpenZFS *might* not compile at all with the new kernel.
```

.. CAUTION::
   

>    **Be sure you don't put anything on that OpenZFS pool that you would need

Reviewer: ```suggestion
   **Be sure you don't put anything on an OpenZFS pool that you would need
```
